pkg/gce-cloud-provider/compute/fake-gce.go

```diff
@@ -422,6 +421,42 @@ func (cloud *FakeCloudProvider) getGlobalSnapshotURI(snapshotName string) string
 		snapshotName)
 }
 
+type FakeBlockingCloudProvider struct {
+	*FakeCloudProvider
+	snapshotToExec chan SnapshotSourceAndTarget
+	readyToExecute chan struct{}
+}
+
+type SnapshotSourceAndTarget struct {
+	Source string
+	Target string
+}
+
+// FakeBlockingCloudProvider's method adds two channels to the CreateSnapshot process in order to provide functionality to finely
+// control the order of execution of volume operations. readyToExecute signals that an CreateSnapshot operation has been called.
+// Then it cycles through the snapshotToExec channel, waiting for permission to actually create the snapshot.
+func (cloud *FakeBlockingCloudProvider) CreateSnapshot(ctx context.Context, volKey *meta.Key, snapshotName string) (*compute.Snapshot, error) {
+	cloud.readyToExecute <- struct{}{}
+	for snapshotToExec := range cloud.snapshotToExec {
+		if snapshotToExec.Source == volKey.String() && snapshotToExec.Target == snapshotName {
+			break
+		} else {
+			cloud.snapshotToExec <- snapshotToExec
```

Comment:
this puts the function into a tight loop when there are items in `snapshotToExec` but they are not for this function. I would like to avoid this, but I've been thinking about it for a while and im honestly not sure how

/cc @misterikkit @msau42

Reply:
we could make this less general and just have 3 channels in the blocking cloud provider, one for each specific snapshot we can execute and just case on those based on which invocation of `CreateSnapshot` we're in. That would get rid of the tight loop as well as the requirement for the exported struct `SnapshotSourceAndTarget`. But that solution is also not ideal as it's definitely pretty janky and requires more channels making it more confusing in a different way.

Either way, don't spend too long thinking about this since it's test code and probably ok as is but its actually kind of an interesting problem. Maybe Jon or Michelle will have some ideas